Lib/test/test_pty.py

```diff
@@ -86,26 +135,66 @@ def handle_sighup(signum, frame):
         # bpo-38547: if the process is the session leader, os.close(master_fd)
         # of "master_fd, slave_name = pty.master_open()" raises SIGHUP
         # signal: just ignore the signal.
+        #
+        # NOTE: the above comment is from an older version of the test;
+        # master_open() is not being used anymore.
         pass
 
-    def test_basic(self):
+    @expectedFailureIfStdinIsTTY
+    def test_openpty(self):
         try:
-            debug("Calling master_open()")
-            master_fd, slave_name = pty.master_open()
-            debug("Got master_fd '%d', slave_name '%s'" %
-                  (master_fd, slave_name))
-            debug("Calling slave_open(%r)" % (slave_name,))
-            slave_fd = pty.slave_open(slave_name)
-            debug("Got slave_fd '%d'" % slave_fd)
+            mode = tty.tcgetattr(pty.STDIN_FILENO)
+        except tty.error:
+            # possible reason: current stdin is not a tty
```

Comment:
I've been away from the code, so this might be a dumb question, but can't we determine if `stdin` is a tty?

Reply:
Sorry for the late reply. That is actually a great question. `tcgetattr()` can fail due to 2 reasons: `errno == EBADF` ( not a valid file descriptor ), `errno == ENOTTY` ( not a tty ). See https://pubs.opengroup.org/onlinepubs/9699919799/. In either case, your stdin is not a tty. In fact, all glibc's `isatty()` does is call `tcgetattr()` to see if it fails; see https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/posix/isatty.c.

I have verified that Linux, FreeBSD, and NetBSD follow this approach for their `isatty()` implementations, while OpenBSD calls `fcntl()+F_ISATTY`. Anyway, the `tcgetattr()` approach works on all the above platforms. I avoid `isatty()`, which is often followed by a separate `tcgetattr()` call, thereby avoiding an extra `tcgetattr()` call [ OCD :D ].

In our code, if `tcgetattr()` fails, then either `pty.STDIN_FILENO`, which is `0`, is a valid fd but not a tty (ENOTTY), or it is not a valid fd at all (EBADF). EBADF will not happen unless something closes `0` [ uncommon scenario ]; for example, by using [ `p close(0)` ] in `gdb`, or if your script itself voluntarily closes `0` for some reason.

The "possible" in the comment  `# possible reason: current stdin is not a tty` is not because of lack of confidence, but because I did not want to discount the (uncommon) chance of an EBADF, in which case `0` is not a valid fd at all [ OCD ]. Anyway, that comment can be safely removed. Do you want me to remove it?

Reply:
I would say update it:

`# possible reason ...` to `# not a tty or bad/closed fd`

and, of course, make that change throughout.

Reply:
Done.